### PR TITLE
Meson

### PR DESCRIPTION
**Pull request by Salamandar**
Add Meson support.
Still missing GObject Introspection support, will come tomorrow.

### CODE REVIEW

**Comment by clefebvre** [comment]
Hi @Salamandar,

This is really interesting for us, Meson looks really cool.

Could you do the following in this PR:

- Rebase the PR (it will result in having .circleci in your branch and trigger a CI build so we can check it builds everywhere).
- Add the "files" directory as is... right now you're listing the one file by naming it, our projects use "files" dirs which have to be installed as is. You can use install_dir in Meson if I'm not mistaken.
- Modify the debian/ packaging, disconnect it from autoconf and connect it to Meson. We build (and so does CI) with dpkg-buildpackage, for the build to use Meson, we need dpkg-buildpackage to call the appropriate meson and ninja commands.

**Comment by Salamandar** [comment]
Hi ! I'm working on it. I did not touch cinnamon-desktop in a while and I've had a lot of experience with Meson since then… So a bit of refactoring is needed. I'll push when all done.

**Comment by Salamandar** [comment]
> You can use install_dir in Meson if I'm not mistaken.
At the time, install_dir would remove all the target directory content. It no longer does that so yes, I can use it.

**Comment by Salamandar** [comment]
> disconnect it from autoconf and connect it to Meson

Do you mean to completely forget about Autotools ? I was keeping the ac/am files for now, but if that's your plan, i'll just remove all autotools files from the repo.

**Comment by Salamandar** [comment]
The debian buildsystem now uses meson, but that installs gir files into `/usr/lib/x86_64-linux-gnu` instead of just `/usr/lib`.
I'll investigate this further.

**Comment by Salamandar** [comment]
Okay, I give up trying to install the correct Meson version (>=0.40) in the CI env.

File: meson.build [hunk 1]
```diff
@@ -0,0 +1,121 @@
+# Meson build file
+
+# https://github.com/linuxmint/cinnamon-desktop
+project('cinnamon-desktop', 'c', version : '3.5.0',
+  meson_version : '>=0.40.0'
```
Comment (leigh123linux) [inline meson.build]: Can the required meson version be reduced to accommodate LMDE3

**Comment by leigh123linux** [comment]
The meson pr has no option for --with-pnp-ids-path and installs a useless file to /usr/share/libcinnamon-desktop/pnp.ids


The --with-pnp-ids-path option is used to specify the path to pnp.ids

```
--with-pnp-ids-path     Specify the path to pnp.ids [default=(internal)]
```
 As the internal file is garbage it should be possible to set it to use hwdata package

```
rpm -qf /usr/share/hwdata/pnp.ids
hwdata-0.308-1.fc28.noarch
```
Linuxmint uses the same path (/usr/share/hwdata/pnp.ids).

https://github.com/linuxmint/cinnamon-desktop/blob/master/debian/rules#L15

**Comment by Salamandar** [comment]
Yeah, there was a TODO left…
Actually I did not have any idea about what's that file, so I just left it as is waiting for your feedback. That worked ;)

File: meson.build (hunk 1 above)
Comment (Salamandar) [reply meson.build]: I tried to use 0.37.0, but 0.40.0 provides `link_whole`.
http://mesonbuild.com/Release-notes-for-0-40-0.html

Comment (leigh123linux) [reply meson.build]: Can you replace link_whole?, this really needs to build on lmde3

Comment (Salamandar) [reply meson.build]: Not really, no. I even waited for Meson to implement this feature.
That's compiler-level, it passes `-Wl,--whole-archive` to the linker. It's needed to expose archive symbols in the shared libraries dynamic symbol table.

By the way, Meson is a really young project with really fast changes. 0.40 was an important release, you can't really *not* ship it with LMDE3.

Comment (Salamandar) [reply meson.build]: Most of the Meson projects will target `0.40.x`. Here : 
* gstreamer https://cgit.freedesktop.org/gstreamer/gstreamer/tree/meson.build
* systemd https://github.com/keszybz/systemd/blob/c4f082a7f87ba888a28525b43921754e1bf29413/meson.build

**Comment by mtwebster** [comment]
Squash please?  Then I'll merge

**Comment by Salamandar** [comment]
Yeah, will do tomorrow !

**Comment by Salamandar** [comment]
I squashed the Meson files modifs, but kept separate the Debian build system modifs and the deletion of Autotools. Tell me if you prefer to have this squashed as well !

**Pull request merged**